src/mlpack/methods/cf/normalization/item_mean_normalization.hpp

```diff
@@ -0,0 +1,157 @@
+ *
+ * // Generate 10 recommendations for all users.
+ * cf.GetRecommendations(10, recommendations);
+ * @endcode
+ */
+class ItemMeanNormalization
+{
+ public:
+  // Empty constructor.
+  ItemMeanNormalization() { }
+
+  /**
+   * Normalize the data by subtracting item mean from each of existing ratings.
+   *
+   * @param data Input dataset in the form of coordinate list.
+   */
+  void Normalize(arma::mat& data)
+  {
+    const size_t itemNum = arma::max(data.row(1)) + 1;
+    itemMean = arma::vec(itemNum, arma::fill::zeros);
+    // Number of ratings for each item.
+    arma::Row<size_t> ratingNum(itemNum, arma::fill::zeros);
+
+    // Sum ratings for each item.
+    data.each_col([&](arma::vec& datapoint) {
+      const size_t item = (size_t) datapoint(1);
+      const double rating = datapoint(2);
+      itemMean(item) += rating;
+      ratingNum(item) += 1;
+    });
+
+    // Calculate item mean and subtract item mean from ratings.
+    // Set item mean to 0 if the item has no rating.
+    for (size_t i = 0; i < itemNum; i++)
+    {
+      if (ratingNum(i) != 0)
+        itemMean(i) /= ratingNum(i);
+    }
+
+    data.each_col([&](arma::vec& datapoint) {
+      const size_t item = (size_t) datapoint(1);
+      datapoint(2) -= itemMean(item);
+      // The algorithm omits rating of zero. If normalized rating equals zero,
+      // it is set to the smallest positive double value.
+      if (datapoint(2) == 0)
+        datapoint(2) = std::numeric_limits<double>::min();
+    });
+  }
+
+  /**
+   * Normalize the data by subtracting item mean from each of existing ratings.
+   *
+   * @param cleanedData Sparse matrix data.
+   */
+  void Normalize(arma::sp_mat& cleanedData)
+  {
+    itemMean = arma::vec(arma::mean(cleanedData, 1));
+
+    arma::sp_mat::iterator it = cleanedData.begin();
+    arma::sp_mat::iterator it_end = cleanedData.end();
+    for (; it != it_end; it++)
+    {
+      *it = *it - itemMean(it.row());
+      // The algorithm omits rating of zero. If normalized rating equals zero,
+      // it is set to the smallest positive double value.
+      if (*it == 0)
+        *it = std::numeric_limits<double>::min();
+    }
+  }
+
+  /**
+   * Denormalize computed rating by adding item mean.
+   *
+   * @param user User ID.
+   * @param item Item ID.
+   * @param rating Computed rating before denormalization.
+   */
+  double Denormalize(const size_t /* user */,
+                     const size_t item,
+                     const double rating) const
+  {
+    return rating + itemMean(item);
+  }
+
+  /**
+   * Denormalize computed rating by adding item mean.
+   *
+   * @param combinations User/Item combinations.
+   * @param predictions Predicted ratings for each user/item combination.
+   */
+  void Denormalize(const arma::Mat<size_t>& combinations,
+                   arma::vec& predictions) const
+  {
+    for (size_t i = 0; i < predictions.n_elem; i++)
+    {
+      const size_t item = combinations(1, i);
```

Comment:
your choice, but I think we could also just write:

```c++
predictions += combinations.submat(...).
```

Reply:
hmm I think we should do something like:
```C++
arma::Col<size_t> items = combinations.row(1);
predictions += itemMean.elem(items);
```
But `X.elem()` takes `arma::uvec` as input and I didn't find a easy conversion from `arma::Col<size_t>` to `arma::uvec`. So unless there is an easy conversion I missed, I think we can stay with the current solution.

Reply:
There is http://arma.sourceforge.net/docs.html#conv_to, but I think since this would cause an extra copy, let's stick with the current implementation, thanks for looking into it.